### DISCUSSION TITLE
Why does explicit interface implementation cause CS0453, where implicit does not?

### DISCUSSION THREAD

**Discussion by ashmind** [dotnet/csharplang#4638] [Q&A] [answered]
Example code:
```csharp
#nullable enable

using System.Collections.Generic;

public interface I1 {
    IEnumerable<T?> M<T>();
}
public class C1 : I1 {
    public IEnumerable<T?> M<T>() => new T[0];
}

public interface I2 {
    IEnumerable<T?> M<T>();
}
public class C2 : I2 {
    // CS0453: The type 'T' must be a non-nullable value type in order 
    // to use it as parameter 'T' in the generic type or method 'Nullable<T>'
    // (and few other errors, but this is the main one I think)
    IEnumerable<T?> I2.M<T>() => new T[0];
}
```

https://sharplab.io/#v2:D4YgdgrgNlCGBGUCmACJYHILAChcAEAGFfARgBYBuXAgZhQEswAXJAJwDNYBjVASVIoA3rhRiSpWgB4AKgH4AfCgCyshQAoAlNRwBfOiQBMKAMKCAXCgHDR4/PTLT5S1TI2aUAXiVgkAdxQZAG1CAF0dfTwce0YWdi5eK2MRHHEJJ0UVNS0Ig3xjE2NLPmTbMUdZTJKAOld3Lx9/QJDw3F0gA===

Why does `C2` cause `CS0453`, while `C1` does not?

**Comment by 333fred** [comment, marked as answer]
This is one of the cases you need a `where T : default`: https://sharplab.io/#v2:D4YgdgrgNlCGBGUCmACJYHILAChcAEAGFfARgBYBuXAgZhQEswAXJAJwDNYBjVASVIoA3rhRiSpWgB4AKgH4AfCgCyshQAoAlNRwBfOiQBMKAMKCAXCgHDR4/PTLT5S1TI2aUAXiVgkAdxQZAG1CAF0dfTwce0YWdi5eK2MRHHEJJ0UVNS0Ig3xjE2NLPmTbMUdZTJKAOld3FD8AC3ZUGRRLABMkLmhmbxRfAOCwiKA=

**Reply by ashmind** [reply]
Thanks! It wasn't obvious at all, but now I get it.

For anyone finding this discussion: the reason why it needs `default` is because explicit implementations (and [overriding members](https://sharplab.io/#v2:D4YgdgrgNlCGBGUCmACJYHILAChcAEAGFfARgBYBuXAgZhQQGcAXAJ1gGNmSAmFAQRQBvXCjEoADqwD2zJFyQATBvBbsuJUrQA8AFQD8APhQBZPYYAUASmo4AvnV4oAwigBcA4aPFTZ8ucrSAG5IrKwAloqoZDoGxma6llYoALzGYEgA7ii6ANqEALq2dkA=)) could do `T?` without `where T: struct` even before C# 9. They inherited constraints from the parent, so it just meant `Nullable<T>` and an expectation for the parent to provide `where T: struct`.

Adding `where T: default` is a pseudo-constraint that makes it clear to the compiler it's the C# 9 `T?` and not historical `Nullable<T>`.

I'll raise a suggestion in Roslyn to improve error experience here.

**Reply by ashmind** [reply]
Raised a Roslyn suggestion: https://github.com/dotnet/roslyn/discussions/52545
